zio/ndjsonio/typeparser.go

```diff
@@ -0,0 +1,326 @@
+	defaultPath string
+	stats       *typeStats
+}
+
+var (
+	ErrDescriptorNotFound = errors.New("descriptor not found")
+	ErrMissingPath        = errors.New("missing path field")
+	ErrBadFormat          = errors.New("bad format")
+)
+
+// Information about the correspondence between the flattened structure
+// of a JSON object and its zng representation (which may include
+// nested record fields). The two descriptors here represent the same data
+// in the same order, flatDescriptor describes the data as it appears in
+// JSON, descriptor describes it as it appears in zng values.
+type typeInfo struct {
+	descriptor *zng.TypeRecord
+	flatDesc   *zng.TypeRecord
+	path       []byte
+}
+
+func getUnsafeDefault(data []byte, defaultValue string, key string) (string, error) {
+	val, err := jsonparser.GetUnsafeString(data, key)
+	if err != nil {
+		// This is always a KeyPathNotFoundError, including if the json was invalid.
+		if defaultValue == "" {
+			return "", jsonparser.KeyPathNotFoundError
+		}
+		return defaultValue, nil
+	}
+	return val, nil
+}
+
+func lookupTypeInfo(zctx *resolver.Context, desc *zng.TypeRecord, path string) *typeInfo {
+	flatCols := zeekio.FlattenColumns(desc.Columns)
+	flatDesc := zctx.LookupTypeRecord(flatCols)
+	info := typeInfo{desc, flatDesc, []byte(path)}
+	return &info
+}
+
+// findTypeInfo returns the typeInfo struct matching an input json
+// object.  If no match is found, an error is returned. If defaultPath
+// is non-zero, it is used as the default _path if the object has no
+// such field. (we could at some point make this a bit more generic by
+// passing in a "defaultFieldValues" map... but not needed now).
+func findTypeInfo(zctx *resolver.Context, jobj []byte, tr typeRules, defaultPath string) (*typeInfo, error) {
+	var fieldName, fieldVal, path string
+	for _, r := range tr.rules {
+		if len(r.Match) != 1 {
+			// This will turn into a panic when we do validation upon loading json types
+			return nil, fmt.Errorf("Rule %v with length %d", r, len(r.Match))
+		}
+
+		var name, val string
+		for n, v := range r.Match {
+			name = n
+			val = v
+		}
+
+		// we keep track of the last field value we extracted
+		// to avoid re-parsing the json object many times to
+		// lift out the same field, as would be the case with
+		// a typical zeek typing config where all rules refer
+		// to the field "_path".
+		if fieldName != name {
+			fieldName = name
+			var err error
+			if name == "_path" {
+				fieldVal, err = getUnsafeDefault(jobj, defaultPath, name)
+				path = fieldVal
+			} else {
+				// jsonparser.Get will return the key even for
+				// some invalid json. For example Get('x{"a":
+				// "b"}', "a") returns "b". This is ok because
+				// these errors will later be caught by ObjectEach.
+				fieldVal, err = jsonparser.GetUnsafeString(jobj, name)
+			}
+			if err != nil {
+				continue
+			}
+		}
+		if val == fieldVal {
+			return lookupTypeInfo(zctx, tr.descriptors[r.Descriptor], path), nil
```

Comment:
Can we cache this info in the descriptor?
Ah, I see there's a comment about this being costly below.  Follow-up optimization?

Reply:
oops, no i left the comment but wanted to do it in this PR... coming

Reply:
done